src/clj/rems/middleware.clj

```diff
@@ -68,13 +68,19 @@
 
 (defn wrap-context [handler]
   (fn [request]
-    (binding [context/*root-path* (calculate-root-path request)
-              context/*roles* (cond-> #{}
-                                context/*user* (set/union (roles/get-roles (getx-user-id))
-                                                          (applications/get-all-application-roles (getx-user-id)))
-                                (:uses-api-key? request) (conj :api-key))]
-      (with-mdc {:roles (str/join " " (sort context/*roles*))}
-        (handler request)))))
+    (let [unavailable-roles (-> request
+                                auth/get-api-key
+                                api-key/unavailable-roles)]
+      (binding [context/*root-path* (calculate-root-path request)
+                context/*roles* (cond-> #{}
```

Comment:
this expression could probably be simplified, perhaps with some intermediate definitions in the let block above

Reply:
Yeah, it's a bit too complicated, I'll do something about it.